src/beanmachine/paic2/src/bm/LowerToFunc.cpp

```diff
@@ -0,0 +1,145 @@
+
+        // We lower "toy.return" directly to "func.return".
+        rewriter.replaceOpWithNewOp<func::ReturnOp>(op);
+        return success();
+    }
+};
+//===----------------------------------------------------------------------===//
+// BMToFunc RewritePatterns: Print operations
+//===----------------------------------------------------------------------===//
+
+struct PrintOpLowering : public OpConversionPattern<bm::PrintWorldOp> {
+    using OpConversionPattern<bm::PrintWorldOp>::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(bm::PrintWorldOp op, OpAdaptor adaptor,
+                                  ConversionPatternRewriter &rewriter) const final {
+        // We don't lower "bm.print_world" in this pass, but we need to update its
+        // operands.
+        rewriter.updateRootInPlace(op,[&] { op->setOperands(adaptor.getOperands()); });
+        return success();
+    }
+};
+//===----------------------------------------------------------------------===//
+// BMToFunc RewritePatterns: Func operations
+//===----------------------------------------------------------------------===//
+
+struct FuncOpLowering : public OpConversionPattern<bm::FuncOp> {
+    using OpConversionPattern<bm::FuncOp>::OpConversionPattern;
+
+    LogicalResult matchAndRewrite(bm::FuncOp op, OpAdaptor adaptor, ConversionPatternRewriter &rewriter) const final {
+        llvm::SmallVector<mlir::Type> types;
+        for(mlir::Type type_input : op.getFunctionType().getInputs()){
+            auto world_ptr = type_input.dyn_cast_or_null<bm::WorldType>();
+            if(world_ptr != nullptr){
+                mlir::ShapedType rankedTensorType = world_ptr.getElementTypes().front();
+                mlir::ShapedType rankedMemRefType = mlir::MemRefType::get(rankedTensorType.getShape(), rankedTensorType.getElementType());
+                types.push_back(rankedMemRefType);
+            } else {
+                types.push_back(type_input);
+            }
+        }
+
+        // TODO: needs to be recursive. Also use TypeConverter?
+        for(auto arg_ptr = op.front().args_begin(); arg_ptr != op.front().args_end(); arg_ptr++){
+            mlir::BlockArgument blockArgument = *arg_ptr;
+            auto world_ptr = blockArgument.getType().dyn_cast_or_null<bm::WorldType>();
+            if(world_ptr != nullptr) {
+                mlir::ShapedType rankedTensorType = world_ptr.getElementTypes().front();
+                mlir::ShapedType rankedMemRefType = mlir::MemRefType::get(rankedTensorType.getShape(), rankedTensorType.getElementType());
+                blockArgument.setType(rankedMemRefType);
+            }
+        }
+        mlir::TypeRange typeRange(types);
+        mlir::FunctionType new_function_type = rewriter.getFunctionType(typeRange, op.getFunctionType().getResults());
+
+        // Create a new function with an updated signature.
+        auto newFuncOp = rewriter.create<mlir::func::FuncOp>(op.getLoc(), op.getName().str(), new_function_type);
```

Comment:
should you be using the TypeConverter instead? https://mlir.llvm.org/docs/DialectConversion/#type-conversion